test/e2e/cluster_logging_gcl.go

```diff
@@ -0,0 +1,292 @@
+/*
+Copyright 2015 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/util/json"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = framework.KubeDescribe("Cluster level logging using GCL", func() {
+	f := framework.NewDefaultFramework("gcl-logging")
+
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("gce")
+	})
+
+	It("should check that logs from pods on all nodes are ingested into GCL", func() {
+		ClusterLevelLoggingWithGcl(f)
+	})
+})
+
+type SynthLoggerEntry struct {
+	NodeNo  int
+	EntryNo int
+}
+
+type LogEntry struct {
+	TextPayload string
+}
+
+const (
+	testLogName             = "test-log"
+	testLogIngestionTimeout = 3 * time.Minute
+)
+
+// ClusterLevelLoggingWithGcl is an end to end test for cluster level logging.
+func ClusterLevelLoggingWithGcl(f *framework.Framework) {
+	By("Checking GCL API readiness")
+	// Write an event to the GCL API to ensure it is working
+	payload := "test_message_" + f.Namespace.Name
+	err := writeEventsToGcl(testLogName, payload)
+	Expect(err).NotTo(HaveOccurred(), "Failed to write event to GCL API")
+
+	// Read an event written earlier to ensure reading from GCL API also works
+	var entries []*LogEntry
+	for start := time.Now(); len(entries) == 0 && time.Since(start) < testLogIngestionTimeout; time.Sleep(5 * time.Second) {
+		entries, err = readFilteredEntriesFromGcl(payload)
+		Expect(err).NotTo(HaveOccurred(), "Failed to read filtered event from GCL API")
+	}
+	Expect(entries).To(HaveLen(1))
+
+	// Wait for the Fluentd pods to enter the running state.
+	By("Checking to make sure the Fluentd pod are running on each healthy node")
+	// Obtain a list of healthy nodes so we can place one synthetic logger on each node.
+	nodes := getHealthyNodes(f)
+	fluentdPods, err := getFluentdPods(f)
+	Expect(err).NotTo(HaveOccurred(), "Failed to obtain fluentd pods")
+	err = waitForFluentdPods(f, nodes, fluentdPods)
+	Expect(err).NotTo(HaveOccurred(), "Failed to wait for fluentd pods entering running state")
+
+	By("Creating dummy loggers")
+	taintName, podNames, err := createSynthLoggers(f, nodes)
```

Comment:
I think there is not much value in running a logging container on every single node. The setup is the same for each node, so if it works on one node it should also work for all nodes. It will make the flow of this test much simpler. WDYT?